### ISSUE TITLE
Spec behavior of ::before and ::after within SVG text

### ISSUE THREAD

**Issue by AmeliaBR** [w3c/svgwg#125]
per this resolution from [12 March 2015](https://www.w3.org/2015/03/12-svg-minutes.html#item04):

> RESOLUTION: That SVG will support ::before and ::after pseudoelements with text content (and possibly other types of replaced content), subject to resolving other text layout issues first

I would like to get this into SVG 2, but Tav is worried about extra complications.

**Comment by tabatkins** [comment]
This one is basically just "finally specify SVG in terms of the CSS box model". Then it falls out of the definition, because only <text> (and a few others) display their children.

**Comment by AmeliaBR** [comment]
It's not quite that simple, since we'd have to cover not only wrapping text but also manually positioned text/tspan and textPath.  So we need to factor in the per-character positioning attributes.  I'd argue that per-character attributes should be assigned *after* inserting the extra text content from the pseudo-element.  Anything else would cause far too much complication.  Basically, layout would treat it as an anonymous `<tspan>` first or last child of the relevant element, with no attributes of its own. 

We'd already agreed only to support text values for `content`, not image.  And we would also ignore any `display` (other than none), `float`, etc, just like we do on a `<tspan>`.

**Comment by tabatkins** [comment]
I agree with all the particulars you list, and I maintain that such behavior falls out naturally from just letting <text> generate a ::before child via the box model. ^_^

Only supporting text content is an interesting restriction. Forcing/ignoring 'display' value, tho, is perfectly fine; there are several examples in CSS of a parent layout mode enforcing certain child layout modes.

**Comment by Tavmjong** [comment]
This gets quite complicated when a the before/after text is unpredictable (say a counter for chapter which could produce one or more characters). How do you deal with the per-character attributes? For the moment we've decided not to do anything and push it to a future spec.

**Comment by tabatkins** [comment]
There's nothing complicated there; the behavior just falls out of the model.  If you have unpredictable ::before text and you need to position stuff per-character, you'll just need to put it in a different element - those two concerns are simply incompatible.

**Comment by AmeliaBR** [comment]
Given the resolution for #200, I don't think there are any remaining obstacles to allowing generated content.  Character counting is done after CSS insertions & deletions.  We already have prose in the chapter with respect to pseudo-elements.  After the changes for #200, it looks like only two other sentences will need to change:

Replace:

> Additionally, the `@font-face` rule must be supported for font selection, as well as the ::first-line and ::first-letter pseudo-elements must be supported on ‘text’ elements.

with

> ...the ::first-line, ::first-letter, ::before, and ::after pseudo-elements must be supported on ‘text’ elements.

(Also, should add a sentence requiring ::selection to be supported in interactive user agents.)
(Also, do we define how `first-line` applies to manually positioned text? Is it all a first line? Or only until you start a new text chunk?)

and replace

> Additionally, the ::before and ::after generated content pseudo-elements must not apply to SVG text elements.

with

> Additionally, any image values for the  `content` property of  `::before` and `::after` generated content pseudo-elements must not apply to SVG text elements.

**Comment by AmeliaBR** [comment]
(and I really, really, really want to one day be able to auto-generate tick mark labels with CSS counters...)

**Comment by nixkuroi** [comment]
Assuming ::before and ::after will be supported in tspan, will the content attribute be fully supported as well?

**Comment by AmeliaBR** [comment]
@nixkuroi When discussing ::before and ::after, we are talking about using `content` to specify text.  However, that's not the same as "fully supported", since `content` can also be used to specify an image.  There is currently no layout model within SVG for an image inside a text element.
